botocore/useragent.py

```diff
@@ -0,0 +1,421 @@
+    """Replaces all not allowed characters in the string with a dash ("-").
+
+    Allowed characters are ASCII alphanumerics and ``!$%&'*+-.^_`|~``. If
+    ``allow_hash`` is ``True``, "#"``" is also allowed.
+
+    :type allow_hash: bool
+    :param allow_hash: Whether "#" is considered an allowed character.
+    """
+    return ''.join(
+        c
+        if c in _USERAGENT_ALLOWED_CHARACTERS or (allow_hash and c == '#')
+        else '-'
+        for c in raw_str
+    )
+
+
+class UserAgentComponent(NamedTuple):
+    """
+    Component of a Botocore User-Agent header string in the standard format.
+
+    Each component consists of a prefix, a name, and a value. In the string
+    representation these are combined in the format ``prefix/name#value``.
+    """
+
+    prefix: str
+    name: str
+    value: Optional[str] = None
+
+    def to_string(self):
+        """Create string like 'prefix/name#value' from a UserAgentComponent."""
+        clean_prefix = sanitize_user_agent_string_component(
+            self.prefix, allow_hash=True
+        )
+        clean_name = sanitize_user_agent_string_component(
+            self.name, allow_hash=False
+        )
+        if self.value is None or self.value == '':
+            return f'{clean_prefix}/{clean_name}'
+        clean_value = sanitize_user_agent_string_component(
+            self.value, allow_hash=True
+        )
+        return f'{clean_prefix}/{clean_name}#{clean_value}'
+
+
+class RawStringUserAgentComponent:
+    """
+    UserAgentComponent interface wrapper around ``str``.
+
+    Use for User-Agent header components that are not constructed from
+    prefix+name+value but instead are provided as strings. No sanitization is
+    performed.
+    """
+
+    def __init__(self, value):
+        self._value = value
+
+    def to_string(self):
+        return self._value
+
+
+class UserAgentString:
+    """
+    Generator for AWS SDK User-Agent header strings.
+
+    The User-Agent header format contains information from session, client, and
+    request context. ``UserAgentString`` provides methods for collecting the
+    information and ``to_string`` for assembling it into the standardized
+    string format.
+
+    Example usage:
+
+        ua_session = UserAgentString.from_environment()
+        ua_session.set_session_config(...)
+        ua_client = ua_session.with_client_config(Config(...))
+        ua_string = ua_request.to_string()
+
+    For testing or when information from all sources is available at the same
+    time, the methods can be chained:
+
+        ua_string = (
+            UserAgentString
+            .from_environment()
+            .set_session_config(...)
+            .with_client_config(Config(...))
+            .to_string()
+        )
+
+    """
+
+    def __init__(
+        self,
+        platform_name,
+        platform_version,
+        python_version,
+        python_implementation,
+        execution_env,
+        crt_version=None,
```

Comment:
It's interesting that the CRT is a top level argument. Do we envision adding more attributes here as other core features are added or do we need something more general to expand metadata/feature info?

Reply:
e19e237 addresses the fact that `crt_version` wasn't actually getting used yet. Including it at all, and as a top level "environment" property, is to maintain [current botocore behavior](https://github.com/boto/botocore/blob/develop/botocore/session.py#L554). The spec doesn't mention the CRT version at all.